### ISSUE TITLE
Linux (Ubuntu 20.04): left alt doesn't work for general key mapping (curly braces and os on)

### ISSUE THREAD

**Issue by amkdev** [rbreaves/kinto#563]
Tried out kinto today and it works nicly, but I noticed that for instance entering curly braces is only possible with the right alt key. 

I'm using the wireless Magic Keyboard, Ubuntu's German (Macintosh) keyboard mapping and of course kinto.

Any hints or ideas? Thanx in advance!

**Comment by rbreaves** [comment]
I use an english keyboard so I am used to just using a dedicated key and shift. I would jump over [here](https://github.com/mooz/xkeysnail) and see if they have any info on it, but this may be a modification to key.py to fix, I am not sure.

https://github.com/mooz/xkeysnail/blob/master/xkeysnail/key.py

**Comment by amkdev** [comment]
Thanks for your reply. Of course I can set my left alt (any alt) entering the third 3rd level in Ubuntu's tweaks ... but than other kinto related mappings doesn't work (for instance app switching)

**Comment by rbreaves** [comment]
There’s an option in tweaks to leave the right Alt key alone for the purpose of various languages but that’s as far as I’ve gone to try & support multi language really.

**Comment by SaschaHeyer** [comment]
I have the same issue `{} and @` on Magic Keyboard only work with the right option key.
Which is not very convenient for programming =D
![MQ052D](https://user-images.githubusercontent.com/1991664/133997714-6bc87b18-5e12-4ef1-ad35-a39aa3256149.jpeg)

Is there a solution to map the same functionality also to the left option key?

**Comment by Scosh** [comment]
hey :) another german speaker here. just wanted to quickly express my gratitude that this project exists at all, as i’m sure it must not have been easy to set up and it’s nice to not feel so alone with such a niche problem and have a group consolidate efforts to try to bring the mac-keyboard philosphy as a viable option to linux. i’m especially thankful for the `cmd+arrow keys` acts as `HOME` and `END`, because i was really missing that!

---

my german keyboard leads to the same issue that @amkdev and @SaschaHeyer describe, the left alt key does not work as expected.

i’m not sure how it is for other languages, but on german keyboards, macOS allows you to use either alt key for special characters hidden on the third input layer. as @SaschaHeyer has alluded to, this is nice for any kind of writing, because all you have to do is start to remember where the characters are located and you can easily input a whole bunch of special characters.

i’m running 42.2 xorg (not wayland) gnome on a relatively fresh arch linux install.

i also tried messing around with gnome-tweaks, but this too seemed to break the rest of the great functionality you already implemented …

i’m a designer, not a programmer, so i don’t know what the gnome settings and gnome tweaks are using to make their keyboard shortcut calls, but i was already running into trouble upstream because before i found `kinto`, i tried to get an approximation of the mac-bindings by checking the `Ctrl is mapped to Alt, Alt to Win` under `Alt and Wind behaviour` in _tweaks_ and then tried varioius combinations of either using tweaks or the standard gnome settings to get the left alt to be an alternate character key, to no avail.

strangely, i _could_ get alt to work like this by checking items like `Any Alt` in `Key to choose 3rd level` but never **in conjunction** with any sort of repositioning of the super/ctrl keys … so this already seems to be an issue with gnome — if i should instead first file an issue with one of their projects, please let me know!

**Comment by joshgoebel** [comment]
I'm a little lost though because I thought (at least for Apple kbs) that we don't remap `Alt` at all - so I'm not sure why we would break any functionality related to the alt keys, unless you're talking about the `Mac/Win` keyboard remapping which does fiddle with alt?

**Comment by joshgoebel** [comment]
And if we are talking `Mac/Win` how do you resolve conflicts... the problem is the keymap turns your left Alt key into the Cmd (RightCtrl) key... 

```
# - Default Mac/Win
Key.LEFT_ALT: Key.RIGHT_CTRL,   # WinMac
```

...so is that what you wanted?  If you want that then now Alt is the command key...  So when you hit `Cmd-L` do you want `AltGr-L` (@) or do you want the actual `Cmd-L` (focus location bar in browser, etc)...

I'm not seeing how you resolve such conflicts when you overload the key like this?  You can configure one of them to win - but you can't have BOTH.

**Comment by joshgoebel** [comment]
>  should have specified that my linux computer where i am using kinto is a thinkpad which of course has the normal windows layout

Ok then yeah you're left with my question about `Cmd-L`... and how to resolve the "two keys on one key" issue.

> alt to work basically like a normal alt until the system encounters a text input field:

Unless this was the answer, and sadly (for X at least) I'm not even sure this is technically possible... we can figure out which window has input focus, but visibility into specific details about that windows UI is either impossible or I worry out-of-scope.

If someone proved it was possible I'd perhaps welcome an exploration into what supporting it could look like (perhaps to develop hooks for plugins)... but not at the cost of TONS of added complexity (which would be my worry).

Why can't you just use the "new" alt key?  Isn't it now moved to exactly where it should be on an Apple keyboard?

**Comment by SebCorbin** [comment]
Hey there, just to add me to the conversation, I'm also really enjoying having macOS shortcuts on Ubuntu with an Apple Keyboard.

Here's my setup:
**Internal keyboard**: Thinkpad T14 (french azerty)
**External keyboard**: Apple USB Keyboard (french azerty)
**Ubuntu keyboard layout** configured as French (Macintosh)
**Kinto Keyboard** configured as Apple

First, I want to say that I swapped some keys in xkeysnail's `key.py` as per #638 

I'm willing to help as much as possible, and so far after reading much of this convo, I don't know what feedback you're missing, so here's a rundown of my tries:

1. with my setup, when I set in Ubuntu keyboard settings the "Alternate Characters Key" to Layout default, right-alt successfully enter third-layer chars and shortcuts work
![image](https://github.com/rbreaves/kinto/assets/645207/f72cf148-9230-4447-99a7-4afda22cce01)

2. when I set the "Alternate Characters Key" to "Left alt" in Ubuntu keyboard settings, I get this layout, and as said before shortcuts based or mapped to Alt-key won't work but I can input third-layer chars
![image](https://github.com/rbreaves/kinto/assets/645207/d8ed5cbf-2927-4fc2-9cd4-c0a16a71ed21)

3. when I set the "Alternate Characters Key" to "default for layout" in Ubuntu keyboard settings and "Any alt" for 3rd level in Gnome tweaks, I get this layout, and shortcuts based or mapped to Alt-key won't work but I can input third-layer chars
![image](https://github.com/rbreaves/kinto/assets/645207/6a0b37a3-22ae-4496-93fb-531781408022)


Now when I read the whole conversation, I think that the issue comes from mapping the whole Super/Alt modifier key to another key globally. Could we prevent this behavior by somehow whitelisting the number row?

**Comment by SebCorbin** [comment]
I think https://github.com/rbreaves/kinto/issues/716 is related too

**Comment by RedBearAK** [comment]
@SebCorbin 

I'm not the Kinto dev, just a minor contributor. But I've been trying to understand this kind of issue for quite a while, and I can add a couple of points of understanding (I hope). 

The Kinto config file only has a modmap exception for the right Alt key, to allow it to be used to get to the Level3 keys on the layout. This is why, when you have this feature enabled in the Kinto tweaks (AltGr on Right Alt), you can get to the additional keys using the key on the right side, but it won't work when you set the key to be on the left. Because it's still being remapped on the other side. 

What is happening with the whole number keys row is the same thing that was happening with A/Q and Z/W, the `key.py` file is based on a standard US layout mapping of key codes to symbols, and that file would need to be modified for all the other keys on French AZERTY that are different from a US QWERTY layout. This includes at least one other letter key (M), several punctuation keys on the right side and the whole number keys row. 

And, importantly, some of the keys on French AZERTY use symbols on the base (non-Shifted) character, that are Shifted characters in the US layout, so they literally aren't named anywhere in the `key.py` file. You have to take the same key code that would be on that key, and change the symbol name that goes with it. 

In some cases this may actually cause the loading of the keymapper config to fail, as it might not be able to find the name of the key that it sees in a keymap. I don't know if you can avoid this by just using aliases in `key.py` to the new symbol names. 

In short, this is going to be a bit complicated to solve, but it should primarily be centered around modifying `key.py` to fully match the keyboard layout you are trying to use. 

What you're talking about with whitelisting the number row... might be possible, but it wouldn't fix the issues with shortcuts on the rest of the keyboard where the keys are also different from the US layout. And you would no longer be able to use any "shortcuts" involving keys on the number row, if they were remapped by the keymapper config onto other shortcuts.
